### ISSUE TITLE
Feedback on proposed custom page handling

### ISSUE THREAD

**Issue by fdmoulin** [perwendel/spark#724]
Three points to consider:

1) CustomErrorPages should not be accessed statically.  Otherwise, multiple service instances in a single application will be forced to share the same set of CustomErrorPages which will cause conflict.  It should be a private member of the Service class with each Service instance having its own instance of CustomErrorPages.  In the case of the Spark class, static accessors are fine since the underlying service instance is encapsulated as a singleton.

2) Static methods in Spark class should be generalized so that the http status code is passed as a parameter.  This will allow a single function to handle all status codes as needed (E.g. 503, 504, etc).  This also has the probable benefit of simpler underlying code.  It reads pretty well too!

Example:

```java
	httpStatus(404, "<html><body>Custom 404</body></html>");

	httpStatus(500, "<html><body>Custom 500</body></html>");

	httpStatus(504, "<html><body>Gateway Timeout</body></html>");

	httpStatus(404, (q, a) -> {
		a.type("application/json");
		return "{\"code\": \"404\",\"message\": \"Not found\"}";
	});
```

3) The CustomErrorPages class should provide accessors to modify and add custom page handlers as needed.  Again, these should not be static.  The Spark class can provide static accessors for convenience, but the underlying Service and its CustomErrorPages should not be limited to static instantiation.  There are valid use cases for the Spark framework where multiple concurrent service instances are needed.

Example:

```java
myService.getCustomPages().setPage(404, "<html><body>Custom 404</body></html>");
myService.getCustomPages().getPage(404);
myService.getCustomPages().removePage(404);
myService.getCustomPages().hasPage(404);
myService.getCustomPages().clearPages();

myService.getCustomPages().setRoute(404, (q, a) -> {
	a.type("application/json");
	return "{\"code\": \"404\",\"message\": \"Not found\"}";
});
myService.getCustomPages().getRoute(404);
myService.getCustomPages().removeRoute(404);
myService.getCustomPages().hasRoute(404);
myService.getCustomPages().clearRoutes();
```

Thanks for the great work on Spark!

**Comment by jakaarl** [comment]
I completely agree on the first point. Disagree on the second, I think "named" methods are much preferable to "generic" ones. Modifying custom pages at any given point doesn't sound like a good idea, IMHO - personally I'd prefer immutability, but a lot of people probably disagree.

On a side note: currently `CustomerErrorPages` is a mutable singleton, with no synchronization -> definitely thread-unsafe.

**Comment by fdmoulin** [comment]
Yes, I agree that named methods are more readable and user friendly.  And this is the key strength of Spark.  Thinking ahead though, I'm assuming that at least some of the other status codes will need to be implemented too (E.g. 400, 401, 403).  This will require a bunch of new named methods (two per status code).  It also makes the API a bit less flexible, but it's a pros and cons thing I suppose.

On the thread-safety topic, what do you think of the use of static calls and members in Spark?  Is this something you've discussed in the past?  Any ideas being thrown around there?  This is one of my pet concerns as I'm using the Spark package in a multi server instance context.  I haven't looked at everything yet but I know of a couple places where it's an issue, like CustomErrorPages or RouteOverview.
